### DISCUSSION TITLE
Conditionally add Response headers based on browser

### DISCUSSION THREAD

**Discussion by Tanmayshetty** [vercel/next.js#36992] [Help] [answered]
I would like to add response headers conditionally based on the browser being used.
This is for React site and not for API.

For Chrome based browsers

> Cross-Origin-Embedder-Policy: credentialless
Cross-Origin-Opener-Policy: same-origin

For Firefox and Safari
> Cross-Origin-Embedder-Policy: require-corp
Cross-Origin-Opener-Policy: same-origin

https://developer.chrome.com/blog/coep-credentialless-origin-trial/
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cross-Origin-Embedder-Policy#browser_compatibility

![image](https://user-images.githubusercontent.com/8419309/168856977-2ad5c4bb-1c25-4006-abfe-e583c79c914c.png)

How can this be achieved?

**Comment by icyJoseph** [comment]
I guess you can use `pages/_middleware.ts`, like this:

```ts
import { NextResponse, type NextRequest } from "next/server";

export function middleware(req: NextRequest) {
  const { pathname } = req.nextUrl;

  if (
    pathname.startsWith("/api") || //  exclude all API routes
    pathname.startsWith("/static") || // exclude static files
    pathname.includes(".") // exclude all files in the public folder
  )
    return NextResponse.next();

  const userAgent = req.ua;

  if (!userAgent) return new Response(null, { status: 400 });

  const browserName = userAgent.browser.name;

  if (!browserName) return new Response(null, { status: 400 });

  const response = NextResponse.next();

  response.headers.set("x-browser-name", browserName);

  return response;
}
```

Does it fit your use case?

**Comment by tar04** [comment]
Hey! Did you found some ways to do this?

**Reply by Tanmayshetty** [reply, marked as answer]
I have added the following code now

``` 
export async function getServerSideProps({ res }) {
  res.setHeader(
    "Cross-Origin-Embedder-Policy",
    uaService.doesBrowserSupportCredentialless() // Browser Check
      ? "credentialless"
      : "require-corp"
  );
})

```
